app/libs/caliper/entity.php

```diff
@@ -60,9 +57,44 @@ public static function session() {
             $duration = $session_end - $session_start;
             $session->setDuration(CaliperSensor::iso8601_duration($duration));
         }
+
+		$extensions = [];
+		if ( array_key_exists( 'HTTP_REFERER', $_SERVER ) ) {
+			$extensions['referer'] = $_SERVER['HTTP_REFERER'];
+		}
+		if ( [] !== $extensions ) {
+			$session->setExtensions( $extensions );
+        }
+
         return $session;
     }
 
+	/**
+	 * Generates a SoftwareApplication entity
+	 */
+	public static function client( $session_id ) {
+		$user_client = ( new SoftwareApplication( ResourceIRI::user_client( $session_id ) ) );
+
+		if ( array_key_exists( 'HTTP_HOST', $_SERVER ) ) {
+			$user_client->setHost( $_SERVER['HTTP_HOST'] );
+		}
+
+		if ( array_key_exists( 'HTTP_USER_AGENT', $_SERVER ) ) {
+			$user_client->setUserAgent( $_SERVER['HTTP_USER_AGENT'] );
+		}
+
+		// get ip address if available.
+		if ( ! empty( $_SERVER['HTTP_CLIENT_IP'] ) ) {
```

Comment:
Should we take the priority as `HTTP_X_FORWARDED_FOR`, then `REMOTE_ADDR`, and put `HTTP_CLIENT_IP` as the last resort?  Users can spoof the  `HTTP_CLIENT_IP` value by sending http header `CLIENT-IP`.

But then, if the server is not behind proxy, users can fool us by setting `X-FORWARDED-FOR` header. (If the server is behind proxy, hopefully the proxy will chain up `HTTP_X_FORWARDED_FOR` correctly). `REMOTE_ADDR` is the best if server is not behind proxy as it is set by the server.  Maybe we should log both `HTTP_X_FORWARDED_FOR` and `REMOTE_ADDR`?

Reply:
interesting read on CakePHP's `clientIp` function vulnerabilities

https://legalhackers.com/advisories/CakePHP-IP-Spoofing-Vulnerability.html

Reply:
Really good point, I'll rearrange them. I didn't put to much through into the order originally (I think I originally just plopped in a list I found of most common possible variables ip addresses can appear in).